### DISCUSSION TITLE
Introducing 'streamer'

### DISCUSSION THREAD

**Discussion by bug-or-feature** [ig-python/ig-markets-api-python-library#313] [Announcements]
**streamer** is a new way to connect with the Streaming API. It provides higher level classes and objects - the idea is that it will be easier to deal with the Streaming API in a real trading app. 

Currently we have support for the CHART part of the Streaming API, via a `ticker` object. A `ticker` is: 
* aware of price changes for an epic
* standalone - can be passed around to different functions
* hides away the threading and implementation details

In future, further support will be added for the other Streaming API functions: MARKET, TRADE, ACCOUNT

To see example use, see
* [sample_ticker.py](https://github.com/ig-python/trading-ig/blob/master/sample/sample_ticker.py) - which displays output much like the existing Streaming sample
* [sample_ticker_rich.py](https://github.com/ig-python/trading-ig/blob/master/sample/sample_ticker_rich.py) - this displays a table showing dynamically updating epic prices. For this one, the [Rich](https://pypi.org/project/rich/) library needs to be installed

**Comment by fxbravo** [comment]
Hi,

First, thank you for you amazing work :)

I have a problem with Lightstreamer (it's OK with Rest API) , and I don't understand how it works and how to stream prices.

I try to run sample-ticker but it doesn't work for me (I'm a newby in python dev)

Can you explain me how to stream price (eg: IX.D.NASDAQ.IFE.IP) in demo IGTrading?

ps: I try with IG companion Streaming, and it works fine.

Thank you very much for your answer.

FX

Here the message from the code:

```
    ticker = self._tickers[epic]
             ~~~~~~~~~~~~~^^^^^^
KeyError: 'IX.D.NASDAQ.IFE.IP'
```

**Comment by galburn** [comment]
Can you please read this page and then give us LOTS more details:-
[https://github.com/ig-python/trading-ig/blob/master/README.rst]

Also, my understanding is that the sample works with Spreadbet accounts, not CFD accounts.  That is probably the best place to start looking.

**Comment by fxbravo** [comment]
Hi galburn,

Thank you for your quick answer !

You were right when you said:  "Also, my understanding is that the sample works with Spreadbet accounts, not CFD accounts. That is probably the best place to start looking."

How can I modify the code in order to stream CFD price ?

Thank you again for your kindness :) 

FIY, here the code with rich table (I do it on PyCharm CE, MacBook Air M2) : 

```
import logging
import time
from trading_ig import IGService, IGStreamService
from trading_ig.config import config
from trading_ig.streamer.manager import StreamingManager
#from sample.sample_utils import crypto_epics  # fx_epics, index_epics, weekend_epics

try:
    from rich.table import Table
    from rich.live import Live
except ImportError:
    print("Rich must be installed for this sample")

FORMAT = "%Y-%m-%d %H:%M:%S"


def main():
    logging.basicConfig(level=logging.WARNING)

    ig_service = IGService(
        config.username,
        config.password,
        config.api_key,
        config.acc_type,
        acc_number=config.acc_number,
    )

    ig = IGStreamService(ig_service)
    ig.create_session(version="3")
    sm = StreamingManager(ig)

    crypto_epics = ["IX.D.NASDAQ.IFE.IP"]
    tickers = []
    for epic in crypto_epics:  # fx_epics, index_epics, crypto_epics
        sm.start_tick_subscription(epic)
        tickers.append(sm.ticker(epic))

    def generate_table(tickers) -> Table:
        table = Table(title="EPIC Prices")
        table.add_column("EPIC")
        table.add_column("Bid", justify="right")
        table.add_column("Offer", justify="right")
        table.add_column("Last traded price", justify="right")
        table.add_column("Last traded volume", justify="right")
        table.add_column("Incremental volume", justify="right")
        table.add_column("Mid open", justify="right")
        table.add_column("Change since open", justify="right")
        table.add_column("Daily % change", justify="right")
        table.add_column("Day high", justify="right")
        table.add_column("Day low", justify="right")
        table.add_column("Timestamp")

        for ticker in tickers:
            table.add_row(
                f"{ticker.epic}",
                f"{ticker.bid:.2f}",
                f"{ticker.offer:.2f}",
                f"{ticker.last_traded_price:.2f}",
                f"{ticker.last_traded_volume}",
                f"{ticker.incr_volume}",
                f"{ticker.day_open_mid:.2f}",
                f"{ticker.day_net_change_mid:.2f}",
                f"{ticker.day_percent_change_mid:.2f}",
                f"{ticker.day_high:.2f}",
                f"{ticker.day_low:.2f}",
                f"{ticker.timestamp.strftime(FORMAT) if ticker.timestamp else ''}",
            )

        return table

    with Live(generate_table(tickers), refresh_per_second=4) as live:
        for _ in range(100):
            time.sleep(0.4)
            live.update(generate_table(tickers))

    for epic in crypto_epics:
        sm.stop_tick_subscription(epic)


if __name__ == "__main__":
    main()
```

**Reply by galburn** [reply]
There are two things to consider:-
1. The streamed prices available to a CFD account are dependent upon IG. Consult them.
2. The developers of trading-ig are not a free education resource.  Support is limited to discovery of bugs within the library and rectification of such. Some of the developers may provide consultancy at a chargable rate, but that is up to them individually.

**Comment by fxbravo** [comment]
Thank you for your answer.

I'll try to find a solution. And if not, I think I will request a dev for this.

Thank you again for your answer

Wish you a great day
